### DISCUSSION TITLE
wsApp.publish or ws.send?

### DISCUSSION THREAD

**Discussion by kvalood** [uNetworking/uWebSockets.js#818] [Q&A] [answered]
Hi all!
What will be more productive when sending 100 messages to 1000 users per second?

```js
wsApp.publish('topic', message)
```

or

```js
clients = new Map() // or object
clietns.forEach(u => {
    u.send(message)
})
```

**Comment by e3dio** [comment]
Publish is much better. Not only will it iterate over 1000 clients faster, but the main feature of Publish is it automatically groups outgoing messages by socket before sending out, if you publish 100 different messages to some subset of 1000 clients it will be 100 times more efficient and faster than looping over and sending each message individually, because it can send all messages to a client at same time

**Reply by kvalood** [reply]
Will it be effective?
- 2000 topics
- 100 posts per topic per second
- 1000 users connected on different topics (from 1 to 10)

**Reply by e3dio** [reply]
Yes that is the ideal case for Publish, it will group all messages together by user in 1 send call, very efficient

**Reply by e3dio** [reply]
Just make sure you minimize message size as much as possible, 100 posts per topic per second sounds like numeric data, make it small with binary protocol

**Reply by kvalood** [reply]
> Just make sure you minimize message size as much as possible, 100 posts per topic per second sounds like numeric data, make it small with binary protocol

Thanks!
such amount of data due to stock quotes. for more than 2000 shares.

**Reply by kvalood** [reply]
@e3dio 

How can one client subscribe twice to the same data source?
Suppose there are several places in the application on one screen where data can be duplicated.

I used to use a `guid` to determine where to output data. 
when subscribing, GUID was set. 
When changing the "topic" on the client side, there was an unsubscribe by GUID and a subscription to a new topic

Example of sending data to the server:
```js
rooms.forEach(ud => {
    let to = ud.subs.filter(i => i.ticker === symbol)
    to.forEach(i => {
        ud.send({                        
            data: [{
                ...
            }],
            guid: i.guid
        })
    })                        
})
```

With the `wsApp.publish('topic/symbol')` method, this cannot be done, because data is sent to everyone at once.

**Reply by e3dio** [reply]
I don't know what you are asking. The client wants duplicate data? They can duplicate it on the client side. You probably need to re-phrase the question

**Reply by kvalood** [reply]
@e3dio I changed the original question a bit. Please read.

**Reply by e3dio** [reply]
Still have no idea what you are asking. Subscribe/Publish by topic is very easy and straightforward, it is not rocket science

**Reply by e3dio** [reply]
> How can one client subscribe twice to the same data source?

This is the only question you asked and it makes zero sense, unless you want duplicate data, which still makes no sense. Need to write a question that makes sense

**Reply by kvalood** [reply]
> This is the only question you asked and it makes zero sense, unless you want duplicate data, which still makes no sense. Need to write a question that makes sense

Duplicate sending data from server to client because client subscribed multiple times

Doesn't make sense, because on the client side, you can receive data once, and output it several times to different places.

**Reply by e3dio** [reply]
> Duplicate sending data from server to client because client subscribed multiple times

You should probably consider re-writing your client code to not subscribe to the same data source multiple times, avoid sending duplicate data. Just a hint ;)

**Comment by uNetworkingAB** [comment, marked as answer]
> Hi all! What will be more productive when sending 100 messages to 1000 users per second?
> 
> ```js
> wsApp.publish('topic', message)
> ```
> 
> or
> 
> ```js
> clients = new Map() // or object
> clietns.forEach(u => {
>     u.send(message)
> })
> ```

wsApp.publish('topic', message) is made for this, so use it. But also make your own benchmarks. The follow up question I don't understand, you use loosely defined, or simply undefined, terminology.
